Bootstrapping - another approach (#88)
Unleash/unleash-client-ruby: README.md

`backup_file` | Filename to store the last known state from the Unleash server. Best to not change this from the default. | N | String | `Dir.tmpdir + "/unleash-#{app_name}-repo.json` |
`logger` | Specify a custom `Logger` class to handle logs for the Unleash client. | N | Class | `Logger.new(STDOUT)` |
`log_level` | Change the log level for the `Logger` class. Constant from `Logger::Severity`. | N | Constant | `Logger::WARN` |
`bootstrap_config` | Bootstrap config on how to loaded data on start-up. This is useful for loading large states on startup without (or before) hitting the network. | N | Unleash::Bootstrap::Configuration | `nil` |

> For in a more in depth look, please see `lib/unleash/configuration.rb`.

Reviewer: ```suggestion
For a more in-depth look, please see `lib/unleash/configuration.rb`.
```

puts "variant color is: #{variant.payload.fetch('color')}"
```

## Bootstrapping

Bootstrap configuration allows the client to be initialized with a predefined set of toggle states. Bootstrapping can be configured by providing a bootstrap configuration when initializing the client.
```ruby
@unleash = Unleash::Client.new(
    url: 'http://unleash.herokuapp.com/api',
    app_name: 'my_ruby_app',
    custom_http_headers: { 'Authorization': '<API token>' },
    bootstrap_config: Unleash::Bootstrap::Configuration.new({
        url: "http://unleash.herokuapp.com/api/client/features",
        url_headers: {'Authorization': '<API token>'}
    })
)
```
The `Bootstrap::Configuration` initializer takes a hash with one of the following options specified:

* `file_path` - An absolute or relative path to a file containing a JSON string of the response body from the Unleash server. This can also be set though the `UNLEASH_BOOTSTRAP_FILE` environment variable.

> * `url` - A url pointing to an Unleash server's features endpoint, the code sample above is illustrative. This can also be set though the `UNLEASH_BOOTSTRAP_URL` environment variable. If using the `url` parameter, a second `url_headers` hash can be passed, which the bootstrapper will pass to the server, see the code sample above for an example. If this option isn't set then the bootstrapper will use the same url headers as the Unleash client.

Reviewer 2: ```suggestion
* `url` - A url pointing to an Unleash server's features endpoint, the code sample above is illustrative. This can also be set though the `UNLEASH_BOOTSTRAP_URL` environment variable. 
* `url_headers` - Headers for the GET http request to the `url` above. Only used if the `url` parameter is also set. If this option isn't set then the bootstrapper will use the same url headers as the Unleash client.
```

* `data` - A raw JSON string as returned by the Unleash server.
* `closure` - A lambda containing custom logic if you need it, an example is provided below.

Example usage:

First saving the toggles locally:
```shell
curl -H 'Authorization: <API token>' -XGET 'http://unleash.herokuapp.com/api' > ./default-toggles.json
```

Now using them on start up:

```ruby

> custom_boostrapper = -> {
>   File.read('./default-toggles.json')
> }

Reviewer 2: I would like to suggest that in the examples we show the `Proc` syntax:

```suggestion
custom_boostrapper = Proc.new {
  File.read('./default-toggles.json')
}
```

Author: Yeah sure! That looks more obvious

```ruby
@unleash = Unleash::Client.new(
    url: 'http://unleash.herokuapp.com/api',
    app_name: 'my_ruby_app',
    custom_http_headers: { 'Authorization': '<API token>' },
    bootstrap_config: Unleash::Bootstrap::Configuration.new({
        url: "http://unleash.herokuapp.com/api/client/features",
        url_headers: {'Authorization': '<API token>'}
    })
)
```
The `Bootstrap::Configuration` initializer takes a hash with one of the following options specified:

* `file_path` - An absolute or relative path to a file containing a JSON string of the response body from the Unleash server. This can also be set though the `UNLEASH_BOOTSTRAP_FILE` environment variable.
* `url` - A url pointing to an Unleash server's features endpoint, the code sample above is illustrative. This can also be set though the `UNLEASH_BOOTSTRAP_URL` environment variable.
* `url_headers` - Headers for the GET http request to the `url` above. Only used if the `url` parameter is also set. If this option isn't set then the bootstrapper will use the same url headers as the Unleash client.
* `data` - A raw JSON string as returned by the Unleash server.

> * `closure` - A lambda containing custom logic if you need it, an example is provided below.

Reviewer 2: May I suggest that we call it `block`?
```suggestion
* `block` - A lambda containing custom logic if you need it, an example is provided below.
```

That's a more common way to call closures in ruby.

Author: Done!

> * `closure` - A lambda containing custom logic if you need it, an example is provided below.

Reviewer 2: We should consider making a table out of this:

```suggestion
Argument | Description | Type | Default Value
-- | -- | -- | -- | --
`file_path` | An absolute or relative path to a file containing a JSON string of the response body from the Unleash server | `String` | ENV['UNLEASH_BOOTSTRAP_FILE'] || nil
`url` | A url pointing to an Unleash server's features endpoint for a GET request. | `String` | ENV['UNLEASH_BOOTSTRAP_FILE'] || nil
`url_headers` | Headers for the GET http request to the `url` above. Only used if the `url` parameter is also set. | `Hash` | Same as the Unleash client
`data`  | A raw JSON string with the features (as returned by the Unleash server). | `string` | nil
`block` | A lambda containing custom logic if you need it, an example is provided below. | `Proc` | nil
```
